bldenv/man/bldquerymerge.man
# @(#)94	1.5  src/bldenv/man/bldquerymerge.man, bldtools, bos412, GOLDA411a 8/3/93 14:02:57
#
# COMPONENT_NAME: (BLDTOOLS) BAI Build Tools
#
# ORIGINS: 27
#
# (C) COPYRIGHT International Business Machines Corp. 1991
# All Rights Reserved
# Licensed Materials - Property of IBM
#
# US Government Users Restricted Rights - Use, duplication or
# disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
#
# NAME: bldquerymerge.man
#
# PURPOSE: man page for bldquerymerge command
#
# EXECUTION ENVIRONMENT: Build process environment
#
#include "bldenv_defines.h"

  bldquerymerge Command

  Purpose

  The bldquerymerge command merges the most recent release makelist files with 
  the release makelist files from previous builds and then executes the 
  bldquerydb command to build an up-to-date version of the bldquery database 
  files.  Once this is complete, bldquerymerge copies all the bldquery database 
  files out to AFS.  For this reason it is required that the user have write 
  permissions to AFS (klog) before running the bldquerymerge command.

  Syntax 

	bldquerymerge [-d]

  Description

  The bldquerymerge command searches the following directories looking for a
  file called "makelist":

	$TOP/PTF/$BLDCYCLE/<release_name>

  If found then this indicates that the release was built during the build 
  cycle specified by the $BLDCYCLE environment variable.  This file is then 
  copied over to the directory:

	$TOP/HISTORY/bldquery/<release_name>

  If a file called "makelist" already exists in this directory then it is 
  replaced.  Once all the release directories have been searched and all the 
  makelist files copied, bldquerymerge invokes the bldquerydb command which 
  creates all the database files required for executing the bldquery command.  
  All the newly created bldquery database files found in the directory,

	$TOP/HISTORY/bldquery

  are then copied to the AFS directory:

	/afs/austin/aix/BLDENV_DEFINE_AFSSUBDIR/HISTORY/bldquery

  This is the default location where the bldquery command looks to find the
  database files necessary for its execution.

  When the bldquerymerge command completes, it posts a status out to the status
  files with a type field of "bldquerymerge" and a status field indicating
  success or failure.

  Flags

	-d - executes bldquerymerge in the delay mode; this causes 
	     bldquerymerge to sleep until midnight before starting the process
	     of building the bldquery database files; if this flag is not set
	     then bldquerymerge begins building the files immediately

  Environment Variables

	TOP		Full pathname of the top of the build tree.  The user
			is prompted for this if it is not set.

	BLDCYCLE	Current build cycle (e.g. 9207).  The user is prompted
			for this if it is not set.

	KLOG_USERID	A valid klog user ID; this will be used by bldquerymerge
			to klog before copying the bldquery database files to 
			AFS.  The bldquerymerge command expects this to be set 
			and does not prompt the user if it is not.  If it
			is not set and there is no active klog token on
			the machine on which bldquerymerge is running, then
			the copy to AFS will fail.

	KLOG_PASSWD	A valid klog password for the user ID specified by the
			KLOG_USERID environment variable; this will be used by 
			bldquerymerge to klog before copying the bldquery 
			database files to AFS.  The bldquerymerge command 
			expects this to be set and does not prompt the user if 
			it is not.  If it is not set and there is no active 
			klog token on the machine on which bldquerymerge is 
			running, then the copy to AFS will fail.

  Examples

	1) bldquerymerge

  Input Files

	$TOP/PTF/$BLDCYCLE/xreflist
	$TOP/PTF/$BLDCYCLE/<release_name>/makelist

  Output Files

	$TOP/HISTORY/bldquery/xreflist
	$TOP/HISTORY/bldquery/<release_name>/makelist

  Related Information

        $BLDENV/usr/bin/bldhostsfile            The bldhostsfile command
                                                used to get name of afs
                                                base directory.
	$BLDENV/usr/bin/bldquery		The bldquery command.
	$BLDENV/usr/bin/bldquerydb		The bldquerydb command.
        $BLDENV/usr/bin/hostsfile.dat           The data file used by
                                                bldhostsfile command.



bldenv/bldtools/bldquerymerge.sh
#! /bin/ksh
# @(#)93	1.13  src/bldenv/bldtools/bldquerymerge.sh, bldtools, bos412, GOLDA411a 1/14/94 13:39:51
#
# COMPONENT_NAME: (BLDTOOLS) BAI Build Tools
#
# FUNCTIONS: bldquerymerge
#
# ORIGINS: 27
#
# (C) COPYRIGHT International Business Machines Corp. 1991
# All Rights Reserved
# Licensed Materials - Property of IBM
#
# US Government Users Restricted Rights - Use, duplication or
# disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
#
# NAME: bldquerymerge
#
# PURPOSE: Queries the status file and generates various formatted reports 
#	   from its contents.  Also allows user to modify/delete entries in
#	   the status file.  Operates from a menu mode or by passing command
#	   line options.
# WARNING: THIS TOOL IS ONLY INTENDED TO BE RUN ON BAI PRODUCTION BUILD 
#          MACHINES.  IT SHOULD NOT BE USED BY AREA BUILDERS.
#
# INPUT: optional command line arguments, BLDCYCLE and STATUS_FILE environment
#	 variables may optionally be initialized before the call.
#
# OUTPUT: Standard output
#
# RETURNS: 0 if successful, 1 if errors
#
# EXECUTION ENVIRONMENT: Build process environment
#

. bldloginit
. bldinitfunc	# define build init functions
. bldkshconst	# define constants
. bldhostsfile
. bldnodenames

function clean_up {
	if [ "$1" = "0" ] 
	then
		bldsetstatus $_TYPE "$T_BLDQUERYMERGE" \
			     $_BLDCYCLE "$BLDCYCLE" \
			     $_BUILDER "$BUILDER" \
			     $_STATUS "$ST_SUCCESS" 
	else
		bldsetstatus $_TYPE "$T_BLDQUERYMERGE" \
			     $_BLDCYCLE "$BLDCYCLE" \
			     $_BUILDER "$BUILDER" \
			     $_STATUS "$ST_FAILURE" 
	fi

	exit $1
}

############################### M A I N ##############################

# call clean_up on interrupts
trap "clean_up 1" HUP INT QUIT TERM

bldinit
chksetbuilder

command=`basename "$0"`			# get command name
LOG=${LOGBLDSTATUS:-"$(bldlogpath)/bldstatus.all"}

LOGOPTS="-c${command} +l -F${LOG}"	# set up log environment

[[ ! -f ${RELEASE_LIST} ]] && \
   log +l -x "Cannot find/open RELEASE_LIST file: ${RELEASE_LIST}"

# checking to see if bldquerymerge is already running; if so then exit now
running=$(ps -ef | grep bldquerymerge | wc | awk -F" *" '{print $2}')
if [ "$running" -gt 2 ]
then
	log ${LOGOPTS} -x "bldquerymerge already running"
fi

delay_flg=""
usage_err=""

while getopts ":d" option
do
      case $option in 
	  d) delay_flg=1;;
          ?) usage_err=1;;
      esac
done

if [ -n "$usage_err" ]
then
	log -x "Usage: ${command} [-d]";
fi

# delaying execution until approximately midnight
if [ -n "$delay_flg" ]
then
	today=$(date | awk -F" " '{ print $3 }')
	while [ "$(date | awk -F" " '{ print $3 }')" = "$today" ]
	do
		sleep 60
	done
fi

# path to local copy of bldquery database
BQDBLOCAL=$(bldhistorypath)/bldquery 
if [ ! -d "$BQDBLOCAL" ]
then
	mkdir ${BQDBLOCAL} || \
		log ${LOGOPTS} -x "could not create directory ${BQDBLOCAL}"
fi

# path to latest raw data to be used in building bldquery data base
BQDBNEW=$(bldglobalpath)

# path to AFS directory where bldquery database files will be stored
bldhostsfile get_afs_base "${TRUE}"
[[ $? -ne 0 ]] && clean_up 1
BQDBGLOBAL=${BQDBGLOBAL:-${HOSTSFILE_AFSBASE}/HISTORY/bldquery}
			
# cleaning up local bldquery database directory
rm -f ${BQDBLOCAL}/* > /dev/null 2>&1
rm -f ${BQDBLOCAL}/LPP/* > /dev/null 2>&1
rm -f ${BQDBLOCAL}/*/SRC > /dev/null 2>&1
rm -f ${BQDBLOCAL}/*/NSRC > /dev/null 2>&1
rm -f ${BQDBLOCAL}/*/SHIP > /dev/null 2>&1
rm -f ${BQDBLOCAL}/*/NSHIP > /dev/null 2>&1
rm -f ${BQDBLOCAL}/*/BLDENV > /dev/null 2>&1

# merging new xreflist and release makelist files into the local bldquery
# database directory

log ${LOGOPTS} -b "merging latest bldquery data with cumulative data"

cp ${BQDBNEW}/xreflist ${BQDBLOCAL}
while read release
do
	if [[ -s "${BQDBNEW}/${release}/makelist"  && \
	      -f "${BQDBNEW}/${release}/DOBLDQUERY" ]]
	then
		if [ ! -d "${BQDBLOCAL}/${release}" ]
		then
			mkdir ${BQDBLOCAL}/${release}
		fi
		cp ${BQDBNEW}/${release}/makelist ${BQDBLOCAL}/${release}
	fi
done < ${RELEASE_LIST}

log ${LOGOPTS} -b "creating bldquery database files"

bldquerydb
if [ "$?" -ne 0 ] 
then
	log ${LOGOPTS} -x "unable to create bldquery database files."
fi

# update AFS w/ the newly created bldquery database files
log ${LOGOPTS} -b "copying bldquery database files to AFS"

if [[ -n "$KLOG_USERID" && -n "$KLOG_PASSWD" ]]
then
	klog $KLOG_USERID -password $KLOG_PASSWD > /dev/null 2>&1
fi

# copy bldquery database files out to AFS
if [ -w "${BQDBGLOBAL}" ]
then
    rm -rf ${BQDBGLOBAL}/*
    cp -r ${BQDBLOCAL}/* ${BQDBGLOBAL}
    if [ "$?" -ne 0 ] 
    then
	log ${LOGOPTS} -x \
	  "unable to copy bldquery database files to ${BQDBGLOBAL}.\n"
    fi
else
    log ${LOGOPTS} -x "directory ${BQDBGLOBAL} is not writeable.\n"
fi

clean_up 0

